src/numerics/include/metaphysicl/dualnumber_surrogate.h

```diff
@@ -0,0 +1,316 @@
+#ifndef METAPHYSICL_DUALNUMBER_SURROGATE_H
+#define METAPHYSICL_DUALNUMBER_SURROGATE_H
+
+#include "metaphysicl/dualnumber_surrogate_decl.h"
+#include "metaphysicl/dualnumber_decl.h"
+
+namespace MetaPhysicL
+{
+
+template <typename T, std::size_t N>
+inline DualNumberSurrogate<T, N>::DualNumberSurrogate(DualNumber<T, NumberArray<N, T>> & dn)
+  : value(dn.value())
+{
+  for (decltype(N) i = 0; i < N; ++i)
+    derivatives[i] = &dn.derivatives()[i];
+}
+
+template <typename T, std::size_t N>
+inline DualNumberSurrogate<T, N>::DualNumberSurrogate(DualNumber<T, NumberArray<N, T>> && dn)
+  : value(dn.value())
+{
+  for (decltype(N) i = 0; i < N; ++i)
+    derivatives[i] = &dn.derivatives()[i];
+}
+
+template <typename T, std::size_t N>
+inline DualNumberSurrogate<T, N>::DualNumberSurrogate(const T & n) : value(n)
+{
+}
+
+template <typename T, std::size_t N>
+template <typename T2, class... Args>
+inline DualNumberSurrogate<T, N>::DualNumberSurrogate(DualNumber<T2, NumberArray<N, T2>> & dn,
+                                                      Args &&... args)
+  : value(dn.value()(std::forward<Args>(args)...))
+{
+  for (decltype(N) di = 0; di < N; ++di)
+    derivatives[di] = &dn.derivatives()[di](std::forward<Args>(args)...);
```

Comment:
This is beautiful.

For the long-term: can you think of any way to avoid hard-coding NumberArray?  I'd love to be able to some day do these tricks with NumberVector, SparseNumberArray, DynamicSparseNumberArray, etc., ideally without copying and pasting code and without stupid macro tricks.

Reply:
I feel like this should mostly just work replacing NumberArray<N, T> with D, since all the classes you named define an `operator[]`?

Reply:
They don't all define an N, though, so you'd have to use ```D::size()```.  Huh, try that?  That would at least get NumberVector working, I think.  This is much less "long-term" than I'd imagined.

For the sparse cases we'd need to use one of the methods for iterating only over elements in the sparsity pattern.  I don't think I've got a properly generic way to do this though, so don't worry about it at all.

Reply:
Alright, I think I was able to totally generalize it, and also my new `NonDuckDualNumber` methods/helper functions through using your `rebind` `structs`